server/src/main/java/keywhiz/inject/StrictGuiceModule.java

```diff
@@ -0,0 +1,11 @@
+package keywhiz.inject;
+
+import com.google.inject.AbstractModule;
+
+public class StrictGuiceModule extends AbstractModule {
+  @Override protected void configure() {
+    binder().disableCircularProxies();
+    binder().requireAtInjectOnConstructors();
+    binder().requireExactBindingAnnotations();
+  }
+}
```

Comment:
Here's what these do:

1. Disallow circular references in our dependency graph
2. Disallow using no-argument constructors for instantiation unless explicitly requested (eliminates an annoying class of subtle bugs)
3. Requires strict matching of binding annotations, addressing a bug in Guice's original design which is kept around for backwards compatibility

Reply:
Yeah, I had a problem where there was a missing binding/logic for loading KeywhizConfig from the yaml file. Because KeywhizConfig is just a data blob with a no-argument constructor Guice happily created a new (empty) one for me, which caused all sorts of bizarre error messages. I've worked extensively with Guice in the past so I knew what to look for, but it's the kind of thing that can leave you running in circles. I decided that it would be best to not have to deal with those sorts of problems entirely at the cost of having to specify 2-3 additional bindings.